### PR TITLE
Use Boto waiters instead of customer _await_status method for RDS Operators

### LINKED ISSUE #27096: Use Boto waiters instead of customer _await_status method for RDS Operators
### Description

Currently some code in RDS Operators use boto waiters and some uses a custom `_await_status`, the former is preferred over the later (waiters are vetted code provided by the boto sdk, they have features like exponential backoff, etc). See [this discussion thread](https://github.com/apache/airflow/pull/27076#discussion_r997325535) for more details/context.

### Use case/motivation

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by hankehly**
Closes: #27096
Related: #27076

### Summary

This PR replaces the status-check/waiter logic in `RdsBaseOperator` and `RdsBaseSensor` with `RdsHook` methods that use boto waiters where possible. All changes are backward compatible.

Please see "Files changed" for detailed information about updates and technical decisions/tradeoffs.

### Todo

- [x] Code changes
- [x] Unit test updates/additions

### CODE REVIEW

File: airflow/providers/amazon/aws/hooks/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/hooks/rds.py:54]: For each of the following resources, I added:
1. A method to get the resource state
2. A method to wait for the resource to reach a certain state

Resources:
- db instance
- db cluster
- db snapshot
- db cluster snapshot
- export task
- event subscription

In the current implementation, `RdsBaseOperator` and `RdsBaseSensor` each have a method that calls the appropriate boto "describe" function based on a string (like "db_snapshot" or "event_subscription").

https://github.com/apache/airflow/blob/64174ce25ae800a38e712aa0bd62a5893ea2ff99/airflow/providers/amazon/aws/operators/rds.py#L49-L66

I considered just moving this logic over to `RdsHook`, but decided to split it up because the existence (and variety) of boto waiters depends on the resource. For example, the db snapshot resource has 3 boto waiters, the db cluster snapshot resource has 2, and the export task resource has none.

One could maybe reduce the amount of code by keep a separate list of resources and their waiters and dynamically choosing at runtime, but I didn't go that far.

File: airflow/providers/amazon/aws/hooks/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/hooks/rds.py]: I wrapped the relevant "resource missing" exceptions in `AirflowNotFoundException` rather than propagating them to the caller. This allows users to catch "resource doesn't exist" exceptions without having to consider the type of resource.

File: airflow/providers/amazon/aws/hooks/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/hooks/rds.py:85]: When boto waiters aren't available, I pass a closure like this to a generic waiter method so that we can reuse waiting logic.

File: airflow/providers/amazon/aws/hooks/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/hooks/rds.py]: Method names and log messages are based off what I saw in other hook classes.

File: airflow/providers/amazon/aws/hooks/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/hooks/rds.py:164]: The default boto waiter delay seems to be 30 for all waiters, but `max_attempts` varies from 40~60. I (arbitrarily) opted to go with 40 everywhere.

File: airflow/providers/amazon/aws/operators/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/operators/rds.py:48]: Replaced with individual `RdsHook` methods.

File: airflow/providers/amazon/aws/operators/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/operators/rds.py:107]: Replaced calls to `_await_status` with `RdsHook` methods throughout this file.

File: airflow/providers/amazon/aws/operators/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/operators/rds.py:184]: You do not need to specify `wait_statuses` when using `RdsHook` waiter methods.

File: airflow/providers/amazon/aws/operators/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/operators/rds.py:247]: This is a feature addition. A boto waiter for `db_snapshot_deleted` exists but wasn't used.

File: airflow/providers/amazon/aws/sensors/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/sensors/rds.py:44]: Basically the same updates as `operators/rds.py`

File: airflow/providers/amazon/aws/sensors/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/sensors/rds.py:159]: `db_type` is a template field, but we were casting it to `RdsDbType` in the constructor, which will raise an exception if we pass a template expression.

File: tests/providers/amazon/aws/hooks/test_rds.py
Comment (hankehly) [inline tests/providers/amazon/aws/hooks/test_rds.py:141]: Added test cases for each method in `RdsHook`

File: tests/providers/amazon/aws/hooks/test_rds.py
Comment (hankehly) [inline tests/providers/amazon/aws/hooks/test_rds.py:148]: Making sure that boto waiters are used when available

File: tests/providers/amazon/aws/hooks/test_rds.py
Comment (hankehly) [inline tests/providers/amazon/aws/hooks/test_rds.py:162]: Making sure custom waiter logic calls the generic waiter method, and the correct boto resource "describe" method

File: tests/providers/amazon/aws/operators/test_rds.py
Comment (hankehly) [inline tests/providers/amazon/aws/operators/test_rds.py:156]: The methods tested here are gone now

File: tests/providers/amazon/aws/operators/test_rds.py
Comment (hankehly) [inline tests/providers/amazon/aws/operators/test_rds.py:209]: `not_called` is not a method. `mock_await_status` is a mock, so the call does not raise an error, but it does not check the intended condition. Replaced with [`assert_not_called`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_not_called) here and below.

File: tests/providers/amazon/aws/operators/test_rds.py
Comment (hankehly) [inline tests/providers/amazon/aws/operators/test_rds.py:232]: Replaced the mocked "wait" method in all RDS operator unit tests.

File: tests/providers/amazon/aws/operators/test_rds.py
Comment (hankehly) [inline tests/providers/amazon/aws/operators/test_rds.py:638]: From the name, this test is supposed to check the "no wait" condition, but `wait_for_completion` was set to `True`. This wasn't caught [for the above reason](https://github.com/apache/airflow/pull/27410/files#r1009685212).

File: tests/providers/amazon/aws/operators/test_rds.py
Comment (hankehly) [inline tests/providers/amazon/aws/operators/test_rds.py:868]: `mock_await_status` wasn't being used (but was probably intended to be) so I added it.

File: tests/providers/amazon/aws/sensors/test_rds.py
Comment (hankehly) [inline tests/providers/amazon/aws/sensors/test_rds.py:122]: The logic tested here is gone now

File: tests/providers/amazon/aws/sensors/test_rds.py
Comment (hankehly) [inline tests/providers/amazon/aws/sensors/test_rds.py:146]: This sensor checks the existence of a snapshot associated with a db instance, but the unit tests didn't create the db instance.

File: airflow/providers/amazon/aws/hooks/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/hooks/rds.py:89]: I see `no overload exists for ...` mypy warnings here because I'm creating the waiter name dynamically. The `if` statement one line above assures we aren't creating a waiter that doesn't exist so I ignore the warning. If someone has a more appropriate solution, please let me know.

File: airflow/providers/amazon/aws/hooks/rds.py
Comment (hankehly) [inline airflow/providers/amazon/aws/hooks/rds.py:54]: For each of the following resources, I added:
1. A method to get the resource state
2. A method to wait for the resource to reach a certain state

Resources:
- db instance
- db cluster
- db snapshot
- db cluster snapshot
- export task
- event subscription

In the current implementation, `RdsBaseOperator` and `RdsBaseSensor` each have a method that calls the appropriate boto "describe" function based on a string (like "db_snapshot" or "event_subscription").

https://github.com/apache/airflow/blob/64174ce25ae800a38e712aa0bd62a5893ea2ff99/airflow/providers/amazon/aws/operators/rds.py#L49-L66

I considered just moving this logic over to `RdsHook`, but decided to split it up because the existence (and variety) of boto waiters depends on the resource. For example, the db snapshot resource has 3 boto waiters, the db cluster snapshot resource has 2, and the export task resource has none.

One could maybe reduce the amount of code by keep a separate list of resources and their waiters and dynamically choosing at runtime, but I didn't go that far.

**Comment by hankehly** [comment]
@o-nikolas @vincbeck @eladkal @ferruzzi
Please review this PR at your earliest convenience.

Also note, I'm having some trouble with the `check-lazy-logging` static check. Previous steps (flynt and pyupgrade) seem to update templated strings in a way that causes the `check-lazy-logging` check to fail. I will continue to tweak it, but please continue with the review in the meantime.

**Comment by o-nikolas** [comment]
@hankehly 
There are more changes in this diff than I expected for the deprecation of `_await_status`, I'll try put aside some time to review this tomorrow :sweat_smile:

**Comment by hankehly** [comment]
> @hankehly There are more changes in this diff than I expected for the deprecation of `_await_status`, I'll try put aside some time to review this tomorrow 😅

Sorry about the larger than expected diff. I thought tweaking the original implementation would result in a lot of if statements. No rush 🙏

File: airflow/providers/amazon/aws/hooks/rds.py
Comment (vincbeck) [inline airflow/providers/amazon/aws/hooks/rds.py]: This is boto3 behavior. Every service specific exception is wrapper inside `ClientError`. [See documentation here](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/error-handling.html#aws-service-exceptions). I would then double check that all the service exception you are catching are really thrown by boto3 (e.g. `DBSnapshotNotFoundFault`, `DBClusterSnapshotNotFoundFault`, ...). They really look service specific exceptions. One proposal as well is is to create a common function we could use across the amazon provider package to get the actual exception. The code `.response["Error"]["Code"]` is widely used

File: airflow/providers/amazon/aws/operators/rds.py
Comment (vincbeck) [reply airflow/providers/amazon/aws/operators/rds.py:247]: Nice! Love that! Could you add unit tests associated to this new feature?

**Review by vincbeck — CHANGES_REQUESTED** [review]
Overall, love the clean up! So much cleaner and more readable. Please double check my comment regarding the exceptions,I think this one is pretty important

File: tests/providers/amazon/aws/operators/test_rds.py
Comment (hankehly) [inline tests/providers/amazon/aws/operators/test_rds.py:391]: Fix: this method was checking the DB _cluster_ id when it should be checking for the DB _instance_ id

Comment (hankehly) [inline tests/providers/amazon/aws/operators/test_rds.py:391]: Fix: `DB_CLUSTER_SNAPSHOT` → `DB_INSTANCE_SNAPSHOT`

File: airflow/providers/amazon/aws/hooks/rds.py
Comment (hankehly) [reply airflow/providers/amazon/aws/hooks/rds.py]: Thanks for mentioning this. I did the following:
* Added another test case for each wait method to check that the appropriate "not found" error is raised
* Replaced all related `try..except` blocks with what the documentation recommends

The documentation states that only _some_ service exceptions are available through dot-notation, but _all_ service exceptions are available through `.response["Error"]["Code"]`.
> Additionally, you can also access some of the dynamic service-side exceptions from the client’s exception property.

.. further down the page:
> Catching exceptions through ClientError and parsing for error codes is still the best way to catch all service-side exceptions and errors.

**Comment by vincbeck** [comment]
Nice! Thanks for the update. LGTM

**Review by vincbeck — APPROVED** [review]

File: airflow/providers/amazon/aws/hooks/rds.py
Comment (vincbeck) [reply airflow/providers/amazon/aws/hooks/rds.py]: LGTM. Thanks for the explanations

**Comment by hankehly** [comment]
The _flynt_ static check modifies log strings (see below) which causes the _check-lazy-logging_ check to fail. If someone knows how to get around this, please let me know.
```diff
- self.log.info("DB cluster snapshot '%s' reached the '%s' state" % (db_instance_id, target_state))
+ self.log.info(f"DB cluster snapshot '{db_instance_id}' reached the '{target_state}' state")
```

**Comment by hankehly** [comment]
I'm getting the following error on the "Build Images / Build CI images 3.7" step. Could someone please re-run it?
```
error: failed to receive status: rpc error: code = Unavailable desc = error reading from server: EOF
```

**Comment by hankehly** [comment]
@o-nikolas @eladkal @ferruzzi
Friendly reminder. This PR has 1 approval and all checks have passed. I would like to merge if possible, or receive some feedback.

**Comment by o-nikolas** [comment]
> Friendly reminder. This PR has 1 approval and all checks have passed. I would like to merge if possible, or receive some feedback.

Hey, still trying to find a block of time to review this one, though do note: I'm not a committer so I wont be able to merge it either way. I also trust @vincbeck's review, if a committer gets to it before I have a chance to review it :+1:

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Woohoo. Nice refactor @hankehly

**Pull request merged**
